numpy/_core/src/common/gil_utils.c

```diff
@@ -34,3 +34,15 @@ npy_gil_error(PyObject *type, const char *format, ...)
     NPY_DISABLE_C_API;
     va_end(va);
 }
+
+// Acquire the GIL before emitting a warning containing a message of
+// the given category and stacklevel.
+NPY_NO_EXPORT int
+npy_gil_warning(PyObject *category, int stacklevel, const char *message)
```

Comment:
I suspect stacklevel is always 1 in anything we do, but also happy to keep it.

Reply:
Yep, I think so as well, but this is intended to be reused elsewhere, so I've kept the same signature as `PyErr_WarnEx`.